src/objects/firmware.py

```diff
@@ -1,56 +1,118 @@
+from typing import Optional
 
 from helperFunctions.hash import get_md5
 from helperFunctions.tag import TagColor
 from objects.file import FileObject
 
 
-class Firmware(FileObject):
+class Firmware(FileObject):  # pylint: disable=too-many-instance-attributes
     '''
-    This class represents a firmware.
-    It is a specialization of :class:`~objects.file.FileObject` that adds a
-    variety of firmware-specific attributes.
-    All constructor arguments are passed to the
-    :class:`~objects.file.FileObject` constructor
+    Uploaded firmware image representation.
+
+    In FACT, we represent an uploaded firmware image as specialized :class:`~objects.file.FileObject` with supplementary meta data.
+    This class is the root of a virtual path tree with all extracted folders and files as branch- and leaf-instances of the :class:`~objects.file.FileObject` class::
+
+                                  ┌────────────┐
+                                  │  Firmware  │
+                                  │(Root Image)│
+                                  └──────┬─────┘
+                                         │
+                         ┌───────────────┼───────────────┐
+                         │               │               │
+                         ▼               ▼               ▼
+                  ┌────────────┐  ┌────────────┐  ┌────────────┐
+                  │    /etc    │  │    /var    │  │    ...     │
+                  │(FileObject)│  │(FileObject)│  │(FileObject)│
+                  └──────┬─────┘  └──────┬─────┘  └────────────┘
+                         │               │
+               ┌─────────┴────┐       ┌──┼──┐
+               │              │       │  │  │
+               ▼              ▼       ▼  ▼  ▼
+        ┌────────────┐ ┌────────────┐   ...
+        │   passwd   │ │    ...     │
+        │(FileObject)│ │(FileObject)│
+        └────────────┘ └────────────┘
+
+
+    For each uploaded firmware, FACT can hold meta data that associates the analyzed image with its corresponding embedded device.
+    This meta data includes the...
+
+    * :attr:`device_class`,
+    * device :attr:`vendor`,
+    * :attr:`device_name`,
+    * firmware :attr:`version`,
+    * firmware :attr:`release_date`, and
+    * image-contained :attr:`part`.
+
+    Additionally, each Firmware can hold user-defined tags that may be used in advanced queries to categorize and filter all firmwares present in the database.
+    It is important to understand that said tags are **separately stored** from the :attr:`objects.file.FileObject.analysis_tags`, which are propagated by analysis plugins.
     '''
 
     def __init__(self, **kwargs):
         super().__init__(**kwargs)
-        #: The name of the device that the firmware runs on
+
+        #: Device name string identifier.
+        #: Can be freely defined during upload.
+        #:
+        #: This attribute is **mandatory** and shall never be `None`.
         self.device_name: str = None
```

Comment:
The optional parameters should use the type `Optional[]` from `typing` (i.e. `Optional[str]` here)